scipy/optimize/minpack.py

```diff
@@ -510,28 +523,34 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         values will all be 1 (if the number of parameters for the function
         can be determined using introspection, otherwise a ValueError
         is raised).
-    sigma : None or M-length sequence, optional
-        If not None, the uncertainties in the ydata array. These are used as
-        weights in the least-squares problem
-        i.e. minimising ``np.sum( ((f(xdata, *popt) - ydata) / sigma)**2 )``
-        If None, the uncertainties are assumed to be 1.
+    sigma : None or M-length sequence or MxM array, optional
+        Determines the uncertainty in `ydata`.
+        If we define measurement errors as
+        ``e = ydata - f(xdata, *popt)``, then
+
+            - If `sigma` is 1-d, it determines standard deviations of errors `e`
+              and the optimized function is ``sum((e / sigma) ** 2)``.
+            - If `sigma` is 2-d, it determines the covariance matrix of errors `e`
+              and the optimized function is ``e.T.dot(inv(sigma)).dot(e)``.
+
+        None (default) is equivalent of 1-d `sigma` filled with ones.
     absolute_sigma : bool, optional
-        If False, `sigma` denotes relative weights of the data points.
-        The returned covariance matrix `pcov` is based on *estimated*
-        errors in the data, and is not affected by the overall
-        magnitude of the values in `sigma`. Only the relative
-        magnitudes of the `sigma` values matter.
-
-        If True, `sigma` describes one standard deviation errors of
-        the input data points. The estimated covariance in `pcov` is
-        based on these values.
+        If False, `sigma` is used in a relative sense. The returned parameter
```

Comment:
Perhaps this explanation is good if you already understand the meaning of `absoulte_sigma=False`. The points I don't like (or don't get):

1. It doesn't clearly state that the final cost function is used to rescale the covariance.
2. It's not super clear what are "scaled errors". We already use scaled errors after introduced sigma, it's just hard to understand that we scale them again.
3. I don't see the benefits of explaining that the cost function will be equal to M - N (people will wonder "why the hell is that and what does it mean?")

I propose my variant, it's just my thoughts, you are welcome to criticize.

```

If False, `sigma` is used in a relative sense. The covariance of errors will be scaled
according to the sample variance of residuals at the found `popt`. Only the relative
magnitudes of the `sigma` values matter.

If True, `sigma` is used in an absolute sense. Use it when the magnitude of errors are
known a priori.

Note, that `absolute_sigma` doesn't affect estimation of `popt`, but only its 
covariance `pcov`.
```